src/rapids_singlecell/preprocessing/_kernels/_bbknn.py

```diff
@@ -0,0 +1,96 @@
+from __future__ import annotations
+
+import cupy as cp
+
+_find_top_k_kernel = r"""
+extern "C" __global__
+void find_top_k_per_row(
+    const float* __restrict__ data,
+    const int* __restrict__ indptr,
+    const int n_rows,
+    const int trim,
+    float* __restrict__ vals) {
+
+    extern __shared__ float shared_memory[];
+
+    int row = blockIdx.x * blockDim.x + threadIdx.x;
+    if (row >= n_rows) return;
+
+    int start = indptr[row];
+    int end = indptr[row + 1];
+    int length = end - start;
+
+    if (length <= trim) {
+        vals[row] = 0.0f;  // Or another default value indicating insufficient elements
+        return;
+    }
+
+    // Each thread has its own top_k array in shared memory
+    int thread_idx = threadIdx.x;
+    int shared_offset = thread_idx * trim;
+    float* top_k = &shared_memory[shared_offset];
+
+    // Initialize top_k with 0
+    for (int i = 0; i < trim; ++i) {
+        top_k[i] = 0;
+    }
+
+    // Process each element in the row
+    for (int idx = start; idx < end; ++idx) {
+        float num = data[idx];
+
+        // Find the smallest value in top_k
+        int min_index = 0;
+        float min_value = top_k[0];
+        for (int i = 1; i < trim; ++i) {
+            if (top_k[i] < min_value) {
+                min_value = top_k[i];
+                min_index = i;
+            }
+        }
+
+        // If current num is larger than the smallest in top_k, replace it
```

Comment:
Ah I get it. It keeps an unsorted array of the *k* largest values encountered so far, by replacing the smallest value each time it finds a value larger than it.

Maybe you could instead maintain a pointer to the smallest value. Then you only need to loop once you find a smaller one, lowering the amortized cost.

Reply:
hmm, if everything reads and writes from the same chunk of shared memory:
- are there race conditions? is it e.g. possible that
  1. core 1 and 2 find that index 5 has the lowest value
  2. core 1 finds a new entry for index 5
  3. core 2 finds a new entry for index 5 and writes it
  4. core 1 writes its entry to index 5, even though it what core 2 wrote is no longer the lowest value
- or will the existence of the shared memory mean that the kernel can’t effectively run in parallel?

I’m pretty sure it’s one or the other since I don’t see any locking or atomics going on.